nexus/reconfigurator/simulation/src/state.rs

```diff
@@ -0,0 +1,510 @@
+    /// only contain one collection.
+    pub fn load_serialized(
+        &mut self,
+        state: UnstableReconfiguratorState,
+        primary_collection_id: Option<CollectionUuid>,
+    ) -> anyhow::Result<LoadResult> {
+        if !self.system.is_empty() {
+            return Err(anyhow!(NonEmptySystemError::new()));
+        }
+
+        let collection_id =
+            get_primary_collection_id(&state, primary_collection_id)?;
+
+        // NOTE: If more error cases are added, ensure that they're checked
+        // before load_serialized_inner is called. This ensures that the system
+        // is not modified if there are errors.
+        let mut res = LoadResultBuilder::default();
+        self.load_serialized_inner(state, collection_id, &mut res);
+
+        Ok(LoadResult {
+            primary_collection_id: collection_id,
+            notices: res.notices,
+            warnings: res.warnings,
+        })
+    }
+
+    // This method MUST be infallible. It should only be called after checking
+    // the invariant: the primary collection ID is valid.
+    fn load_serialized_inner(
+        &mut self,
+        state: UnstableReconfiguratorState,
+        primary_collection_id: CollectionUuid,
+        res: &mut LoadResultBuilder,
+    ) {
+        res.notices.push(format!(
+            "using collection {} as source of sled inventory data",
+            primary_collection_id,
+        ));
+        let primary_collection = state
+            .collections
+            .iter()
+            .find(|c| c.id == primary_collection_id)
+            .expect("invariant: primary collection ID is valid");
+
+        for (sled_id, sled_details) in
+            state.planning_input.all_sleds(SledFilter::Commissioned)
+        {
+            let Some(inventory_sled_agent) =
+                primary_collection.sled_agents.get(&sled_id)
+            else {
+                res.warnings.push(format!(
+                    "sled {}: skipped (no inventory found for sled agent in \
+                     collection {}",
+                    sled_id, primary_collection_id
+                ));
+                continue;
+            };
+
+            let inventory_sp = inventory_sled_agent
+                .baseboard_id
+                .as_ref()
+                .and_then(|baseboard_id| {
+                    let inv_sp = primary_collection.sps.get(baseboard_id);
+                    let inv_rot = primary_collection.rots.get(baseboard_id);
+                    if let (Some(inv_sp), Some(inv_rot)) = (inv_sp, inv_rot) {
+                        Some(SledHwInventory {
+                            baseboard_id: &baseboard_id,
+                            sp: inv_sp,
+                            rot: inv_rot,
+                        })
+                    } else {
+                        None
+                    }
+                });
+
+            // XXX: Should this error ever happen? The only case where it
```

Comment:
I'd maybe move this comment down into the `match`? On first read I wasn't sure what error it was talking about, since there's no `?` and the method is documented as infallible.

Reply:
Good idea! I've done that and combined it with the existing comment below.